### DISCUSSION TITLE
do we have a (data)type instruction?

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#116] [Q&A] [answered]
I've started with the ABS function and must distinguish between integer and float (addressing ILT/FLT & ISUB/FSUB). I haven't found a type instruction to distinguish between them. Is there any other way?

**Comment by adesutherland** [comment]
We come to an important unresolved (certainly unimplemented) discussion point.

In REXX Level B its easy, you need .INT.ABS(), .FLOAT.ABS(), or .DECIMAL.ABS(). The compiler will automatically change ABS(-1) to .INT.ABS(-1) (ignoring that the optimiser will ideally just change it to "1"). We may or may may not need  .STRING.ABS() because if it is a string value the compiler type safety will automatically convert to float (or decimal?) and back - but I am uncertain on this point.

Real Rexx B code might look like
```
a = -1.2 /* Float */
say a.abs() /* says 1.2 - Which I think REXX programmers will find sensible and comprehensible */
```
But that is not what we are working on yet - so back to Classic REXX

For REXX Level C. There are two situations.

1. The compiler is SURE that the value must by an INT or FLOAT etc. In which case the compiler will call the right version of ABS (e.g. int.abs() - we need a version for each type). Note: We have not yet got a way of getting the '.' in names in REXX so just use "intabs" in the short term.

2. The compiler cannot be sure of the type.

See the section (I have copied following) from https://github.com/adesutherland/CREXX/wiki/REXX-Assembler-(RXAS)

> Dynamic Type Instructions
>
> The compiler will be able to manage which registers have what values most of the time but there will be certain dynamic situations where the value type is not known (we will know details of this better with experience but let's predict the future problem now!)
> 
> To handle this the compiler can use the registers type flag,
> 
> gettp reg - This gets the type flag value (combined 0=unset, 1=string, 2=decimal, 4=float, 8=integer, 16=object)

**NOTE: We can't do decimal - that is a mistake. Decimal is an object, and we will come to that later.**

> 
> getandtp reg, reg - This gets the flag value and does a bitwise AND on it so that a particular type can be checked
> 
> settp reg - sets the flag
> 
> setortp reg - sets the flag by doing a bitwise OR on it so that an additional particular type can be set
> 
> The following instructions are the only ones that actually use the flag. Other than these it is up to the compiler to generate the right code to process the type flag.
> 
> These convert the value to integer, string etc. from whatever is the most appropriate value determined by the type flag
> 
> toi reg
> tos reg
> tof reg
> tod reg **(no we can't do this)**
> 

So the idea is the compiler sets the type flags **if it thinks they are needed** manually, and the "dynamic versions of" rxas/rexx code set / reads them as need be. Nothing is automatically done by the vm **except supporting toi, tos, tof**.

So we need a dynamic.abs() which uses "gettp" to determine the register type and acts appropriately. It should also "settp" of the return value.

The trouble is that none of these instructions exist yet (note to PM: we need some microcode!) 

However in interpreter/rxvmvars.h line 85 plus there are some helper functions (but these are also not quite right, all we need to do is read/set  v->status.all_type_flags

**Reply by Peter-Jacob** [reply]
ok, is this implemented yet? I thought I take the easy one first (ABS), but it looks this was an error.

**Reply by adesutherland** [reply]
ABS is not implemented  - but is that what you mean?

**Reply by Peter-Jacob** [reply]
no, I meant is the above-mentioned syntax implemented:  
```
a = -1.2 /* Float */
say a.abs() 
```
I have prepared already an ABS function, but it seems that x.abs() is not compiled accordingly.

**Comment by rvjansen** [comment]
Real Rexx B code might look like
```
a = -1.2 /* Float */
say a.abs() /* says 1.2 - Which I think REXX programmers will find sensible and comprehensible */
```
This, by the way, is valid NetRexx. As is '1.2'.abs()
We have the same dilemma as Java had, with introducing native types. I heard that in hindsight, the developers would have liked to have chosen for boxed types only, but it was too late.

I think the above approach is very sensible. I'll put the instructions in the planning. We will probably need those before finishing up the bif library, if we want to proceed with maximum speed; there will be more situations apart from ABS().

**Comment by adesutherland** [comment, marked as answer]
I am adding the following instructions needed for optional arguments which also cover the dynamic instructions. Obviously I have changed the instruction design slightly!
```
gettp - gets the register type flag (op1 = op2.typeflag)
settp - sets the register type flag (op1.typeflag = op2)
setortp - or the register type flag (op1.typeflag = op1.typeflag || op2)
brtpt - if op2.typeflag true then goto op1
brtpandt - if op2.typeflag && op3 true then goto op1
```

typeflag is a 64bit integer and the compiler / programmer can use it as they want to. However it is envisaged that it will be treated as a bit field. For REXX Level B the compiler sets it to a non-zero value if an optional argument has a value, or zero if the argument is not provided.

**Comment by rvjansen** [comment]
Hi Peter,

no, sorry, my fault: that page indicates the NetRexx syntax. We are just implementing Classic Rexx syntax at the moment.

René

**Reply by Peter-Jacob** [reply]
ok, then let me re-phrase my question. What does the rexx compiler create for this statement:
```
x=-2
say abs(x) 
y=-3.14
say abs(y) 
```
my first thought was maybe something as IABS(x) for integer values or FABS() for floats. Alternatively, I can put the variable type distinction into the ABS function itself, this was where we started the discussion.

**Comment by rvjansen** [comment]
If your rexx is;

/* rexx */
options levelb
x=-2
say abs(x)
y=-3.14
say abs(y)

abs: procedure = .int
arg numer = .string

then the generated code is:

/*
 * cREXX COMPILER VERSION : cREXX I0134
 * SOURCE                 : peter2
 * BUILT                  : 2021-10-06 19:01:52
 */

.globals=0

main() .locals=4 .expose=peter2.main
   * Line 3: say abs(x)
   * Line 3: abs(x)
   load r1,1
   load r2,"-2"
   call r3,abs(),r1
   itos r3
   say r3
   * Line 5: say abs(y)
   * Line 5: abs(y)
   load r2,1
   load r3,"-3.14"
   call r1,abs(),r2
   itos r1
   say r1
   * Line 5:
   ret 0

The great examples are in https://github.com/adesutherland/CREXX/wiki/REXX-Assembler-(RXAS) <https://github.com/adesutherland/CREXX/wiki/REXX-Assembler-(RXAS)>
Very usable source example is tests/levelbfunc.rexx

René.
